src/services/obs/backend.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::io::Result;
+use std::{fmt::Debug, sync::Arc};
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+use log::info;
+use once_cell::sync::Lazy;
+use regex::Regex;
+use reqsign::services::huaweicloud::obs::Signer;
+
+use crate::error::{other, BackendError};
+use crate::http_util::{
+    new_request_build_error, new_request_send_error, new_request_sign_error, parse_error_response,
+    percent_encode_path,
+};
+use crate::ops::BytesRange;
+use crate::Scheme;
+use crate::{
+    http_util::HttpClient,
+    ops::{OpCreate, OpDelete, OpList, OpRead, OpStat, OpWrite},
+    Accessor, AccessorMetadata, BytesReader, BytesWriter, DirStreamer, ObjectMetadata,
+};
+
+use super::error::parse_error;
+
+static OBS_DEFAULT_ENDPOINT_PATTERN: Lazy<Regex> =
+    Lazy::new(|| Regex::new("^(https://)?obs.([a-z]+-[a-z]+-[0-9]+).myhuaweicloud.com$").unwrap());
```

Comment:
I don't like the idea of introducing `regex`.

How about check `starts_with("obs.")` and `ends_with(".myhuaweicloud.com")` on `url.domain()`?

Reply:
fixed